src/node_report.cc

```diff
@@ -0,0 +1,943 @@
+  MallocedBuffer<char> buffer(buffer_size);
+  int rc = -1;
+  while (rc != 0 && buffer_size < 1024 * 1024) {
+    rc = loadquery(L_GETINFO, buffer.data, buffer_size);
+    if (rc == 0) {
+      break;
+    }
+    buffer_size += buffer_inc;
+    buffer = MallocedBuffer<char>(buffer_size);
+  }
+  if (rc == 0) {
+    char* buf = buffer.data;
+    ld_info* cur_info = nullptr;
+    writer->json_arraystart("sharedObjects");
+    do {
+      std::ostringstream str;
+      cur_info = reinterpret_cast<ld_info*>(buf);
+      char* member_name = cur_info->ldinfo_filename +
+          strlen(cur_info->ldinfo_filename) + 1;
+      if (*member_name != '\0') {
+        str << cur_info->ldinfo_filename << "(" << member_name << ")";
+        writer->json_element(str.str());
+        str.str("");
+      } else {
+        writer->json_element(cur_info->ldinfo_filename);
+      }
+      buf += cur_info->ldinfo_next;
+    } while (cur_info->ldinfo_next != 0);
+    writer->json_arrayend();
+  }
+#elif __sun
+  Link_map* p;
+
+  writer->json_arraystart("sharedObjects");
+  if (dlinfo(RTLD_SELF, RTLD_DI_LINKMAP, &p) != -1) {
+    for (Link_map* l = p; l != nullptr; l = l->l_next) {
+      writer->json_element(l->l_name);
+    }
+    writer->json_arrayend();
+  }
+
+#elif _WIN32
+  // Windows implementation - get a handle to the process.
+  HANDLE process_handle = OpenProcess(PROCESS_QUERY_INFORMATION|PROCESS_VM_READ,
+                                      FALSE, GetCurrentProcessId());
+  if (process_handle == nullptr) {
+    return;
+  }
+  // Get a list of all the modules in this process
+  DWORD size_1 = 0, size_2 = 0;
+  // First call to get the size of module array needed
+  if (EnumProcessModules(process_handle, nullptr, 0, &size_1)) {
+    MallocedBuffer<HMODULE> modules(size_1);
+
+    // Second call to populate the module array
+    if (EnumProcessModules(process_handle, modules.data, size_1, &size_2)) {
+      writer->json_arraystart("sharedObjects");
+      for (int i = 0;
+           i < (size_1 / sizeof(HMODULE)) && i < (size_2 / sizeof(HMODULE));
+           i++) {
+        TCHAR module_name[MAX_PATH];
+        // Obtain and report the full pathname for each module
+        if (GetModuleFileNameEx(process_handle,
+                                modules.data[i],
+                                module_name,
+                                sizeof(module_name) / sizeof(TCHAR))) {
+          writer->json_element(module_name);
+        }
+      }
+      writer->json_arrayend();
+    }
+  }
+
+  // Release the handle to the process.
+  CloseHandle(process_handle);
+#endif
+}
+
+// Obtain and report the node and subcomponent version strings.
+static void PrintComponentVersions(JSONWriter* writer) {
+  std::stringstream buf;
+
+  writer->json_objectstart("componentVersions");
+
+  buf << HTTP_PARSER_VERSION_MAJOR << ".";
+  buf << HTTP_PARSER_VERSION_MINOR << ".";
+  buf << HTTP_PARSER_VERSION_PATCH;
+  writer->json_keyvalue("http_parser", buf.str());
```

Comment:
Possible to use the new `NODE_VERSIONS_KEYS` macro? ... like in https://github.com/nodejs/node/blob/master/src/node.cc#L181-L185

Reply:
use the generic macro instead of individual keys